### PR TITLE
Fixed issue #2294.

### LINKED ISSUE #2294: FullNode BTC network memory leak [bug] [sp]
@dangershony reported an issue with Bitcoin network synching using fullnode that smells of memory leak

here the screenshot

image

Dan guess it's a background task not disposing, I'll investigate the issue reporting here the findings

### PR DESCRIPTION
**Pull request by MithrilMan**
Fixes #2294 

Prevent enqueue too many BlockDownload that may cause the process to Fill up all the available computer RAM.

### CODE REVIEW

File: src/Stratis.Bitcoin/Utilities/AverageCalculator.cs [hunk 1]
```diff
@@ -17,6 +17,14 @@ public class AverageCalculator
         /// <summary>Samples used in calculation of the average value.</summary>
         private CircularArray<double> samples;
 
+        /// <summary>
+        /// Gets a value indicating whether this instance is empty.
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if this instance is empty; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsEmpty { get { return this.samples.Count == 0; } }
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Utilities/AverageCalculator.cs]: No need for that- blocks can't be smaller than header size so in GetAverageBlockSizeBytes you can check average to be less or equal to 80. 
Also this change breaks 2 BP tests

Comment (MithrilMan) [reply src/Stratis.Bitcoin/Utilities/AverageCalculator.cs]: moving the bugfix logic into ProcessDownloadQueueLocked method

File: src/Stratis.Bitcoin/Consensus/ConsensusManager.cs [hunk 2]
```diff
@@ -1100,7 +1100,15 @@ private void ProcessDownloadQueueLocked()
                     return;
                 }
 
+                // To fix issue https://github.com/stratisproject/StratisBitcoinFullNode/issues/2294#issue-364513736
+                // if there are no samples, assume the worst scenario (you are going to donwload full blocks).
                 long avgSize = (long)this.blockPuller.GetAverageBlockSizeBytes();
+                if (avgSize == 0)
+                {
+                    // TODO: move MaxBlockSize on a common place.
+                    avgSize = Block.MaxBlockSize;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: No this is wrong it should come from `Netowrk.Consensus`

Comment (noescape00) [reply src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: Yes. We have this value copypasted all over. Should just refactor the code to make sure it's in one place. 

Probably next thing @MithrilMan can do is to find all block size values and making sure we use only one from Network.Consensus

Comment (noescape00) [reply src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: That's why there is a TODO. 
I think @MithrilMan already discussed that in fn channel

Comment (MithrilMan) [reply src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: yes, sparse delcarations all around
I can try to replace all Block.MaxBlockSize to `network.Consensus.Options.MaxBlockBaseSize` if that's the current property (what does `Base` refers to?)

Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: yes @MithrilMan please replace to `network.Consensus.Options.MaxBlockBaseSize` and get rid of all other places with a block size.

Comment (MithrilMan) [reply src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: actually there are 16 references, would be quick to just update last one and then looking to change them all (not just Block. but even other const around, like Transaction.MAX_BLOCK_SIZE)

Comment (MithrilMan) [reply src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: #2375

**Pull request merged**